pkg/ingester/stream_rate_calculator.go

```diff
@@ -87,33 +93,26 @@ func (c *StreamRateCalculator) updateRates() {
 	c.allRates = rates
 }
 
-func (c *StreamRateCalculator) Rates() []*logproto.StreamRate {
+func (c *StreamRateCalculator) Rates() []logproto.StreamRate {
 	c.rateLock.RLock()
 	defer c.rateLock.RUnlock()
 
-	rates := make([]*logproto.StreamRate, 0, len(c.allRates))
-	for _, r := range c.allRates {
-		rates = append(rates, &logproto.StreamRate{
-			StreamHash:        r.StreamHash,
-			StreamHashNoShard: r.StreamHashNoShard,
-			Rate:              r.Rate,
-		})
-	}
-
-	return rates
+	return c.allRates
 }
 
-func (c *StreamRateCalculator) Record(streamHash, streamHashNoShard uint64, bytes int64) {
+func (c *StreamRateCalculator) Record(tenant string, streamHash, streamHashNoShard uint64, bytes int) {
 	i := streamHash & uint64(c.size-1)
 
 	c.locks[i].Lock()
 	defer c.locks[i].Unlock()
 
-	streamRate := c.samples[i]
+	streamRate := c.samples[i][tenant]
```

Comment:
double checking: any chance of samples[i] not having the tenant key on it? if so, maybe you'll have to check `if streamRate, ok := c.samples[i][tenant]; ok {...`

Reply:
If there's not tenant key, we'll just get an empty `logproto.StreamRate`. We ensure that all the maps are initialized here: https://github.com/grafana/loki/blob/04f0a57d1e654b52f943a920fa322064c1093b60/pkg/ingester/stream_rate_calculator.go#L46-L49